### ISSUE TITLE
Feature request - Support a Datalist/Typeahead filter

### ISSUE THREAD

**Issue by jmzagorski** [ghiscoding/aurelia-slickgrid#17]
**I'm submitting a feature request**

* **Browser:**
all 

* **Language:**
TypeScript

**Current behavior:**
Not implemented

* **What is the motivation / use case for changing the behavior?**
Many users prefer the datalist/typeahead feature because it is faster than scrolling through a select list. https://github.com/ghiscoding/aurelia-slickgrid/issues/9 is where the conversation got started, and below is a summary. 

1. The html `datalist` is not supported in Safari and partially support in IE
2. Libraries to consider: [purejs-datalist-polyfill](https://github.com/Fyrd/purejs-datalist-polyfill), [jquery datalist polyfill](https://github.com/CSS-Tricks/Relevant-Dropdowns) and [typeahead](https://twitter.github.io/typeahead.js/). The drawback to this is adding another dependency in this library.

**Comment by ghiscoding** [comment]
Safari is becoming the new IE, in terms of following the standards... just add to say it lol

**Comment by jmzagorski** [comment]
yup, when you said that in the original pull request I started rethinking this as a supported feature. Even as I was typing this issue I was wondering if I:

1. Should let the user add their own custom datalist filter so they can pick their favorite library instead of the library getting another dependency
2. Or just support the plain HTML datalist and specify in WIKI it is currently not supported in safari.

Thoughts?

**Comment by ghiscoding** [comment]
Yeah but the thing is that Safari is too big to ignore, I mean it's probably also Safari on mobile and yeah too big to ignore. On my current App project for work, it will be used by internal users, Sales Rep, External Users and I don't have control over the latter. Also if our internal users and Sales Rep uses an iPad or even an iPhone, then I guess it falls on the Safari club. 

Here's a good reference [Can I use - datalist](https://caniuse.com/#feat=datalist), it's not implemented in Safari/iOS/OperaMini though we shouldn't care about the latter. 

What I always wondered is the following though
- is there any way in Aurelia (probably more of a TypeScript question though) to be able create a Service which requires certain lib but if the user does not use that Service (neither does he have the lib installed) then not include it in the Build?

**Comment by ghiscoding** [comment]
BTW, I wouldn't use Typeahead.js as a Filter, it's too big of a lib to use it for a simple Filter. a Polyfill for datalist is much smaller.

As a side note, I had already started some work in my other `Aurelia-Bootstrap-Plugins` for Typeahead.js,  I had it working with Remote, here's the [branch](https://github.com/ghiscoding/Aurelia-Bootstrap-Plugins/tree/feature/typeahead), I just never released it because I didn't know how to include it in CLI (RequireJS) and started working on Aurelia-Slickgrid, so the Typeahead went on the side.

**Comment by jmzagorski** [comment]
Thanks. I will look into the polyfills. 

> is there any way in Aurelia (probably more of a TypeScript question though) to be able create a Service which requires certain lib but if the user does not use that Service (neither does he have the lib installed) then not include it in the Build?

As far as I know Aurelia does not and that might be more of a bundlers (webpack) job. For example, I do not use the `mulitple-select` dependency by not specifying it in my `boot.js` or `webpack.config` file. Not sure if that helps.

**Comment by ghiscoding** [comment]
You're not using the `multiple-select`, is that something you are already doing? 
We could perhaps include a note somewhere on how to exclude it from the build, maybe in the HOWTO or Filter wiki. I never heard of the `boot.js` and I'd like to know what line you modify in the `webpack.config`. 

By the way for wiki, we can use the ` ```diff ` and add `+` or `-` in front of each lines to show which one to add/remove. I'm a very visual person, so I like to see those when it makes sense

**Comment by jmzagorski** [comment]
I was using the `multiple-select` just as an example since it is a dependency (I do plan on using it in my application, I just have not implemented it yet).

The `boot.js` is the same as `main.js` with Aurelia, I just called it a different file in my application. For my donet core application i use two separate `webpack.config` files with the `DLLPlugin` and that is how I make the `multiple-select` an optional dependency. My `webpack.config.vendor.js` file, which has your library, looks like this:

```javascript
const path = require('path');
const webpack = require('webpack');
const ExtractTextPlugin = require('extract-text-webpack-plugin');
const extractCSS = new ExtractTextPlugin('vendor.css');
const { AureliaPlugin } = require('aurelia-webpack-plugin');

module.exports = ({ prod } = {}) => {
  const isDevBuild = !prod;

  return [{
    stats: {
      modules: false
    },
    resolve: {
      extensions: ['.js']
    },
    // needed for jquery plugins
    externals: {
      jquery: 'jQuery'
    },
    module: {
      rules: [{
        // embed small images and fonts as Data Urls and larger ones as files:
        test: /\.(png|gif|jpg|jpeg|cur)$/i,
        loader: 'url-loader',
        options: {
          limit: 8192
        }
      }, {
        test: /\.woff2(\?v=[0-9]\.[0-9]\.[0-9])?$/i,
        loader: 'url-loader',
        options: {
          limit: 10000,
          mimetype: 'application/font-woff2'
        }
      }, {
        test: /\.woff(\?v=[0-9]\.[0-9]\.[0-9])?$/i,
        loader: 'url-loader',
        options: {
          limit: 10000,
          mimetype: 'application/font-woff'
        }
      }, {
        // load these fonts normally, as files:
        test: /\.(ttf|eot|svg|otf)(\?v=[0-9]\.[0-9]\.[0-9])?$/,
        loader: 'file-loader'
      }, {
        test: /\.html$/i,
        use: 'html-loader'
      }, {
        test: /\.css(\?|$)/,
        use: ExtractTextPlugin.extract({
          fallback: 'style-loader',
          use: [
            { loader: isDevBuild ? 'css-loader' : 'css-loader?minimize' },
            {
              loader: 'postcss-loader',
              options: { plugins: () => [require('autoprefixer')({ browsers: ['last 2 versions'] })]}
            }
          ]
        })
      }]
    },
    entry: {
      // FIXME: in karma testing if i include commented out aurelia libraries then i get
      // missing module errors
      vendor: [
        '@progress/kendo-ui',
        '@progress/kendo-ui/css/web/kendo.material.css',
        '@progress/kendo-ui/css/web/kendo.common.css',
        'arrive',
        'aurelia-dialog',
        'aurelia-fetch-client',
        'aurelia-logging-console',
        'aurelia-router',
        'aurelia-value-converters',
        'aurelia-history-browser',
        'aurelia-pal-browser',
        'aurelia-polyfills',
        'aurelia-event-aggregator',
        // uncomment these if using OR 
        // include a variable with a flag that is set at build time
        //'aurelia-slickgrid/aurelia-slickgrid/assets/lib/multiple-select/multiple-select.js',
       // 'aurelia-slickgrid/aurelia-slickgrid/assets/lib/multiple-select/multiple-select.css',
        'aurelia-slickgrid',
        // TODO adding these in makes Tests fail
        //'aurelia-bootstrapper',
        //'aurelia-kendoui-bridge',
        //'aurelia-templating',
        //'aurelia-templating-binding',
        //'aurelia-templating-resources',
        //'aurelia-templating-router',
        'bluebird',
        'bootstrap',
        'bootstrap/dist/css/bootstrap.css',
        'bootstrap-material-design/dist/css/bootstrap-material-design.css',
        'bootstrap-material-design/dist/css/ripples.css',
        'bootstrap-material-design/dist/js/material.js',
        'bootstrap-material-design/dist/js/ripples.js',
        'font-awesome/css/font-awesome.css',
        'jquery',
        'lodash.isequal',
        'moment',
        'redux',
        'redux-form',
        'redux-actions',
        'redux-logger',
        'redux-saga',
        'reselect',
        'slickgrid'
      ]
    },
    output: {
      path: path.join(__dirname, 'wwwroot', 'dist'),
      publicPath: '',
      filename: '[name].js',
      library: '[name]_[hash]'
    },
    plugins: [
      // for redux
      new webpack.DefinePlugin({
        'process.env': {
          NODE_ENV: JSON.stringify(prod ? 'production' : '')
        }
      }),
      extractCSS,
      new webpack.ProvidePlugin({
        $: 'jquery',
        jQuery: 'jquery',
        'window.jQuery': 'jquery',
        'window.$': 'jquery',
        moment: 'moment',
        Promise: 'bluebird'
      }),
      new webpack.DllPlugin({
        path: path.join(__dirname, 'wwwroot', 'dist', '[name]-manifest.json'),
        name: '[name]_[hash]'
      }),
      // needed to aurelia plugins
      new AureliaPlugin({
        aureliaApp: undefined,
        dist: 'native-modules'
      })
    ].concat(isDevBuild ? [] : [
      new webpack.optimize.UglifyJsPlugin({
        compress: {
          warnings: false
        }
      })
    ])
  }];
};
```

**Comment by ghiscoding** [comment]
I usually never touch that portion, but am I right in assuming that if you start to fill in the `entry.vendor` of your `webpack.config`, you then need to make sure to include any new ones? For example, yesterday I released the Export to File which requires a new 3rd party lib `text-encoding-utf-8`, you would have to go in your config and add it in there else it won't be part of the build at all... right?

**Comment by jmzagorski** [comment]
I do not know, I might not have to include `text-encoding-utf8` in the vendor entry, but I can test it out later. Normally, when I install another library with `npm install` then yes I need to make sure I include it in vendor entry

**Comment by ghiscoding** [comment]
Thanks I created a [Wiki - Exclude 3rd party libs](/ghiscoding/aurelia-slickgrid/wiki/Exclude-3rd-Party-libs) with pieces of your `webpack.config`. Let me know if it make sense or not. I haven't test any of that, so it's up in the air for now but I added it to the wiki index anyway.

You could also add it to your list of TODO in your documentation issue.

**Comment by ghiscoding** [comment]
BTW, I saw `bootstrap-material-design` in your list, if ever want to make the grid a bit more material design I'm open to that. I actually thought about that in the past and I assume it should be easy enough. We can just copy the [slick-bootstrap.scss](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/styles/slick-bootstrap.scss) and/or change a few of the [_variables.scss](https://github.com/ghiscoding/aurelia-slickgrid/blob/master/aurelia-slickgrid/src/aurelia-slickgrid/styles/_variables.scss)

**Comment by jmzagorski** [comment]
Good documentation. I will add that to the TODO list to make sure I test the `text-encoding-utf8` to see if I need to add that.

I will let you know if we ever need the grid to be more like material design. Right now my users don't mind the current theme.

**Comment by ghiscoding** [comment]
That's great because I put lot of efforts in that theme, and the rest of the library too I guess lol.

**Comment by jmzagorski** [comment]
The only thing i do not like about the datalist is that it filters the dropdown when the user types so they cannot see the other values after it is selected. In my app I used a input on top of a select list, but it really does not look nice. If we were to implement this would you want the list filtered like the native implementation or like my app?

**Native**
![2018-06-28_08-05-18](https://user-images.githubusercontent.com/8404957/42033265-54b3c424-7aaa-11e8-96ad-dc60a4e5fc00.gif)

**MyApp**
![2018-06-28_08-09-16](https://user-images.githubusercontent.com/8404957/42033336-9437fb6a-7aaa-11e8-8d83-fd24583717a6.gif)

**Comment by ghiscoding** [comment]
I think the best is to have an "x" in the corner so that the user can clear the filter. As shown in a typeahead below. This is actually an input which I added a Font Awesome icon that has an onClick event on it.

![typeahead](https://user-images.githubusercontent.com/643976/42035770-b7052558-7ab1-11e8-8c47-a0774d23310c.png)

**Comment by ghiscoding** [comment]
@jmzagorski 
I felt on this [Grid Demo](http://6pac.github.io/SlickGrid/examples/example-autocomplete-editor.html) that has an Autocomplete Editor (last column). It's not exactly the same as what you are trying to do here but it might help. 

They use `jQuery UI` which is a also in our lib since it's a dependency and it doesn't look too hard to implement.

```javascript
var countryList = [
  "Afghanistan",
  "Albania",
  "Algeria",
  // ...
];

var columns = [
    {
      id: "CountryOfOrigin", name: "Country Of Origin", field: "country", 
      editor: AutoCompleteEditor, dataSource: countryList
    }
  ];


function AutoCompleteEditor(args) {
  // ...

  this.init = function () {
      $input = $("<INPUT id='tags' class='editor-text' />");
      $input.appendTo(args.container);
      $input.focus().select();
      $input.autocomplete({
          source: args.column.dataSource
      });
  };
```

Seems interesting, this might be reusable for this ticket. 

![2018-09-18_16-48-58](https://user-images.githubusercontent.com/643976/45715678-ca4c7e80-bb62-11e8-8032-f386c3196440.gif)

**Comment by jmzagorski** [comment]
I think this will be fine. Using a dependency already in the library is a good idea so as not to introduce more. This should work for me based on the jquery UI API because you can provide a label/value array object which would make it act like a select list that accepts new values

Thinking out loud here, and a way to reduce dependency in the core of this library and make the testing surface smaller, I wonder if it would be beneficial to eventually refactor out custom editors into their own plugins that rely on external dependencies...for example have an aurelia-slickgrid-jquery plugin or something like that. Just a thought.

**Comment by ghiscoding** [comment]
Right I didn't look at jQuery UI API yet but I would probably use an array of objects as well, I almost never use plain text because that is not what I usually save in DB.

Separating into external smaller lib might be doable to reduce footprint, however at this point I'm more interested into building the entire grid functionalities than improving performance or footprint. It is an interesting concept though, so I'm not excluding it for the future.

**Comment by ghiscoding** [comment]
closed by PR #122
